### PR TITLE
Add CLI command for simulating end device uplinks

### LINKED ISSUE #3144: Simulate AS uplinks in the CLI
#### Summary

With #2919 the AS supports simulating uplinks. Add support in the CLI.

#### Why do we need this?

- Allow users to easily test their payload formatters.
- Feature parity with v2
- Feature parity with the Console (after #2234)

#### What is already there? What do you see now?

HTTP API, usage as described in https://github.com/TheThingsNetwork/lorawan-stack/pull/2919#issue-449575569

#### What is missing? What do you want to see?

Simulate uplinks using the CLI. Also documentation.

Bonus points for receiving the uplink from the AS as well (which contains decoded fields) and printing it to stdout.

#### Environment

3.9.1

#### How do you propose to implement this?

Add a new command `ttn-lw-cli end-devices simulate-up`

#### How do you propose to test this?

Test locally

#### Can you do this yourself and submit a Pull Request?

Yes, but adding to backlog for now. Let's move this higher up if there is interest (:+1:) from the community.

### PR DESCRIPTION
**Pull request by neoaggelos**
#### Summary

Closes #3144 

#### Changes

- Add a new `ttn-lw-cli end-devices simulate-uplink` command, accepting arguments for the uplink message.

#### Testing

- `ttn-lw-cli login`
- `ttn-lw-cli end-devices simulate-uplink app-id dev-id --uplink-message.f-port 13 --uplink-message.frm-payload 01010202`
- Uplink message is received from the Application Server, events are sent, integrations are working as they should.

##### Regressions

There should not be any.

#### Checklist

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [X] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [X] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [X] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Review by johanstokking — COMMENTED** [review]
We already have a top-level `simulate` command which can be confusing with this one.

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 1]
```diff
@@ -1120,6 +1122,38 @@ To generate a QR code for multiple end devices:
...
+		RunE: func(cmd *cobra.Command, args []string) error {
+			devID, err := getEndDeviceID(cmd.Flags(), args, true)
+			if err != nil {
+				return err
+			}
+			if !config.ApplicationServerEnabled {
+				return errApplicationServerDisabled.New()
+			}
+
+			// fport, _ := cmd.Flags().GetUint32("f-port")
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/end_devices.go]: Remove

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 2]
```diff
@@ -1120,6 +1122,38 @@ To generate a QR code for multiple end devices:
...
+			}
+			if err := util.SetFields(up.Up, applicationUplinkFlags); err != nil {
+				return err
+			}
+
+			cc, err := api.Dial(ctx, config.ApplicationServerGRPCAddress)
+			if err != nil {
+				return err
+			}
+			up.EndDeviceIdentifiers = *devID
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/end_devices.go]: This looks obsolete

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 3]
```diff
@@ -1120,6 +1122,38 @@ To generate a QR code for multiple end devices:
...
+			if err != nil {
+				return err
+			}
+			if !config.ApplicationServerEnabled {
+				return errApplicationServerDisabled.New()
+			}
+
+			// fport, _ := cmd.Flags().GetUint32("f-port")
+			up := &ttnpb.ApplicationUp{
+				EndDeviceIdentifiers: *devID,
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/end_devices.go]: With this

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 4]
```diff
@@ -1120,6 +1122,38 @@ To generate a QR code for multiple end devices:
...
+				Simulated:            true,
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/end_devices.go]: Is this necessary when calling `SimulateUplink()`?

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 5]
```diff
@@ -45,6 +45,7 @@ var (
 	endDeviceFlattenPaths    = []string{"provisioning_data"}
 	endDevicePictureFlags    = &pflag.FlagSet{}
 	endDeviceLocationFlags   = util.FieldFlags(&ttnpb.Location{}, "location")
+	applicationUplinkFlags   = util.FieldFlags(&ttnpb.ApplicationUplink{}, "uplink_message")
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/end_devices.go]: Can we collapse this, so that it becomes `--frm-payload` ?

**Review by johanstokking — CHANGES_REQUESTED** [review]

**Review by htdvisser — COMMENTED** [review]
Nothing to add to @johanstokking's review.

**Review by johanstokking — APPROVED** [review]

**Comment by rvolosatovs** [comment]
Why not put this under `simulate application-uplink` or something similar?
We already have `ttn-lw-cli simulate` namespace and we have `ttn-lw-cli simulate uplink` as well. The suggested naming scheme is extremely confusing

**Comment by neoaggelos** [comment]
> We already have a top-level simulate command which can be confusing with this one.

> Why not put this under simulate application-uplink or something similar?

The existing `ttn-lw-cli simulate` command is hidden, marked expiremental and largely undocumented, hence why I avoided using it.

**Comment by htdvisser** [comment]
@rvolosatovs makes a good point. I think we can make this one `simulate application-uplink` and maybe even rename the other one to `simulate gateway-uplink` (it's experimental and hidden, no compatibility expected there).

File: CHANGELOG.md [hunk 6]
```diff
@@ -11,8 +11,12 @@ For details about compatibility between different releases, see the **Commitment
 
 ### Added
 
+- Support for sending end device uplinks using the CLI (see `ttn-lw-cli simulate application-uplink` command).
+
 ### Changed
 
+- Renamed expiremental command `ttn-lw-cli simulate uplink` to `ttn-lw-cli simulate gateway-uplink`.
```
Comment (htdvisser) [inline CHANGELOG.md]: "experimental" instead of "expiremental"

File: cmd/ttn-lw-cli/commands/simulate.go [hunk 7]
```diff
@@ -417,12 +429,12 @@ var (
...
 		Use:     "simulate",
 		Aliases: []string{"sim"},
-		Short:   "Simulation commands (EXPERIMENTAL)",
-		Hidden:  true,
+		Short:   "Simulation commands",
 	}
 	simulateJoinRequestCommand = &cobra.Command{
-		Use:   "join-request",
-		Short: "Simulate a join request (EXPERIMENTAL)",
+		Use:    "join-request",
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/simulate.go]: 1. Maybe this should also have the `gateway-` prefix
2. I think the description should be more clear about the fact that this "simulates a join request from an end device, sent through a simulated gateway connection"

File: cmd/ttn-lw-cli/commands/simulate.go [hunk 8]
```diff
@@ -502,8 +514,9 @@ var (
 	}
 
 	simulateDataUplinkCommand = &cobra.Command{
-		Use:   "uplink",
-		Short: "Simulate a data uplink (EXPERIMENTAL)",
+		Use:    "gateway-uplink",
+		Short:  "Simulate a gateway uplink message (EXPERIMENTAL)",
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/simulate.go]: I think the description should be more clear about the fact that this "simulates an uplink message from an end device, sent through a simulated gateway connection". I think "gateway uplink message" alone is not clear enough.

File: cmd/ttn-lw-cli/commands/simulate.go [hunk 9]
```diff
@@ -637,13 +650,43 @@ var (
 			)
 		},
 	}
+	simulateApplicationUplinkCommand = &cobra.Command{
+		Use:   "application-uplink [application-id] [device-id]",
+		Short: "Simulate application uplink message from an end device",
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/simulate.go]: I think the description should be more clear about the fact that this "simulates an application-layer uplink message from an end device, sent directly to the application server"

**Pull request merged**